notation/verifier.go

```diff
@@ -0,0 +1,159 @@
+	// Required.
+	TrustStore truststore.X509TrustStore
+
+	// PluginManager manages the plugins installed for Notation verifier. It
+	// should implement the plugin.Manager interface:
+	// https://github.com/notaryproject/notation-go/blob/release-1.3/plugin/manager.go#L33
+	// Optional.
+	PluginManager plugin.Manager
+}
+
+// Verifier is a ratify.Verifier implementation that verifies Notation
+// signatures.
+type Verifier struct {
+	name        string
+	sigVerifier notation.Verifier
+}
+
+// NewVerifier creates a new Notation verifier.
+func NewVerifier(opts *NewVerifierOptions) (ratify.Verifier, error) {
+	verifier, err := notationVerifier.New(opts.TrustPolicyDoc, opts.TrustStore, opts.PluginManager)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create notation verifier: %w", err)
+	}
+
+	return &Verifier{
+		name:        opts.Name,
+		sigVerifier: verifier,
+	}, nil
+}
+
+// Name returns the name of the verifier.
+func (v *Verifier) Name() string {
+	return v.name
+}
+
+// Type returns the type of the verifier which is always `notation`.
+func (v *Verifier) Type() string {
+	return notationVerifierType
+}
+
+// Verifiable returns true if the artifact is a Notation signature.
+func (v *Verifier) Verifiable(artifact ocispec.Descriptor) bool {
+	return artifact.ArtifactType == notationSignatureArtifactType && artifact.MediaType == ocispec.MediaTypeImageManifest
+}
+
+// Verify verifies the Notation signature.
+func (v *Verifier) Verify(ctx context.Context, opts *ratify.VerifyOptions) (*ratify.VerificationResult, error) {
+	subjectRef, err := registry.ParseReference(opts.Subject)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse subject reference: %w", err)
+	}
+
+	signatureDesc, err := v.getSignatureBlobDesc(ctx, opts.Store, subjectRef, opts.SubjectDescriptor)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get signature blob descriptor: %w", err)
+	}
+
+	signatureBlob, err := opts.Store.FetchBlobContent(ctx, subjectRef.Repository, signatureDesc)
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch signature blob: %w", err)
+	}
+
+	result := &ratify.VerificationResult{
+		Verifier: v,
+	}
+	if outcome, err := v.verifySignature(ctx, opts.Subject, opts.SubjectDescriptor, signatureBlob, signatureDesc.MediaType); err != nil {
+		result.Err = err
+	} else {
+		cert := outcome.EnvelopeContent.SignerInfo.CertificateChain[0]
+		result.Detail = map[string]string{
+			"Issuer": cert.Issuer.String(),
+			"SN":     cert.Subject.String(),
+		}
+		result.Description = "Notation signature verification succeeded"
+	}
+
+	return result, nil
+}
+
+func (v *Verifier) getSignatureBlobDesc(ctx context.Context, store ratify.Store, artifactRef registry.Reference, artifactDesc ocispec.Descriptor) (ocispec.Descriptor, error) {
+	if artifactDesc.Size > maxManifestSizeLimit {
+		return ocispec.Descriptor{}, fmt.Errorf("signature manifest too large: %d bytes", artifactDesc.Size)
+	}
```

Comment:
This check seems unnecessary. This check should be in the `Store` instance.

Reply:
The manifests are pretty standard while the blobs are per verifier.

Reply:
Synced offline, move all size validation to Store implementation.